sympy/solvers/solveset.py

```diff
@@ -1754,6 +1781,148 @@ def _is_logarithmic(f, symbol):
     return rv
 
 
+def _is_bivariate(f, symbol):
+    r"""
+    Return True if the equation has two generators, else False.
+    """
+    poly = f.as_poly()
+    gens = _filtered_gens(poly, symbol)
+
+    return len(gens) == 2
+
+
+def _solve_as_bivariate(lhs, rhs, symbol, domain):
+    r"""
+    Helper solver to solve equations of bivariate form.
+    """
+    result = ConditionSet(symbol, Eq(lhs - rhs, 0), domain)
+
+    poly = lhs.as_poly()
+    gens = _filtered_gens(poly, symbol)
+    gpu = bivariate_type(lhs - rhs, *gens)
+    if gpu is not None:
+        g, p, u = gpu
+        if g != lhs:
+            usol = _solveset(p, u, domain)
+            if not isinstance(usol, ConditionSet):
+                result = [_solveset(g - us, symbol, domain) for us in usol]
+                res = EmptySet
+                for result1 in result:
+                    res += result1
+                result = res
+    return result
+
+
+def _is_lambert(f):
+    r"""
+    Return True if the equation is of Lambert type, else False.
+    Equations containing more than two variables (in proper format) involving
+    `Pow`, `HyperbolicFunction`,`TrigonometricFunction`, `log` or `exp`
+    terms are currently treated as Lambert types.
+    """
+    term_factors = list(_term_factors(f.expand()))
+
+    lambert_condition1 = len(Add.make_args(f)) >= 2 and len([arg for arg in term_factors if arg.has(Symbol)]) >= 2
+    lambert_condition2 = any(isinstance(arg, (log, HyperbolicFunction, TrigonometricFunction)) and arg.has(Symbol) for arg in term_factors)
+    lambert_condition3 = any(isinstance(arg, (Pow, exp)) and arg.has(Symbol) for arg in term_factors if (arg.as_base_exp()[1]).has(Symbol))
+
+    return lambert_condition1 and lambert_condition2 or lambert_condition3
```

Comment:
These lines are too long. It would be clearer if you wrote this as:
```
if len(...) > 2:
    if an(...):
        return True
    if any(...):
        return True
return False
```
The variable name `lambert_condition1` tells us nothing about what it means. Should be `has_two_terms` or something?

Why are these particular conditions being checked? Why check `Add.make_args` and `term_factors` as well?

Do you think that someone looking at this code would understand what it is doing?

The code should be written more clearly and if it is still not obvious what it is doing then there should be comments explaining what it does. Also the docstring should match what the function actually does.

Reply:
> Should be has_two_terms or something?
> Why are these particular conditions being checked? Why check Add.make_args and term_factors as well?

yes but not just only 2 variables...also to have at least 2 operands and also two variables in each term_factor...

Reply:
> Should be has_two_terms or something?
> Why are these particular conditions being checked? Why check Add.make_args and term_factors as well?

yes but not just only 2 variables...also to have at least 2 operands and also two variables in each term_factor...

Reply:
> Do you think that someone looking at this code would understand what it is doing?

Oh! sorry for this, I'll try better next time!

Reply:
And the reason for using `lambert_condition3` and `lambert_condition2` 
In, `exp and pow `terms the exponent should have a symbol whereas this is not valid for  `log, HyperbolicFunction, TrigonometricFunction`

Reply:
And also `len(eq.free_symbols)` can't be used because 

```
>>> _is_lambert(3*log(x) - x*log(3))
{x} f.free_symbols
False
```

Reply:
```suggestion
    term_factors = list(_term_factors(f.expand()))
    # total number of symbols in equation
    no_of_symbols = len([arg for arg in term_factors if arg.has(Symbol)])
    no_of_operands = len(Add.make_args(f))

    if no_of_operands >= 2 and no_of_symbols >= 2 :
        # `log, HyperbolicFunction, TrigonometricFunction` should have symbols
        if any(isinstance(arg, (log, HyperbolicFunction, TrigonometricFunction))\
            for arg in term_factors if arg.has(Symbol)):
            return True
        # here, `Pow, exp` exponent should have symbols
        elif any(isinstance(arg, (Pow, exp)) \
            for arg in term_factors if (arg.as_base_exp()[1]).has(Symbol)):
            return True
    return False
```